src/poetry/mixology/term.py

```diff
@@ -151,21 +151,45 @@ def difference(self, other: Term) -> Term | None:
         """
         return self.intersect(other.inverse)
 
+    @staticmethod
+    def _is_direct_reference(dependency: Dependency) -> bool:
+        return dependency.source_type in ["directory", "file", "url", "git"]
+
     def _compatible_dependency(self, other: Dependency) -> bool:
         compatible: bool = (
             self.dependency.is_root
             or other.is_root
             or other.is_same_package_as(self.dependency)
         )
+
+        if (
+            not compatible
+            and not self._is_direct_reference(self.dependency)
+            and self._is_direct_reference(other)
+        ):
+            # we do this here to indicate direct reference dependencies are compatible
+            # with NVR dependencies
+            return bool(
+                self.dependency.complete_name == other.complete_name
+                and self.dependency.constraint.allows_all(other.constraint)
+            )
+
         return compatible
```

Comment:
That is a place where `is_same_package_as()` is expected to be non-commutative (which seems really bad design to me). We should refactor it further so it still works when we make `is_same_package_as()` commutative in the future. Assuming, we don't need the constraint check here:

By the way, what does NVR stand for?

```suggestion
        return (
            self.dependency.is_root
            or other.is_root
            or other.is_same_package_as(self.dependency)
            or (
                # we do this here to indicate direct reference dependencies are
                # compatible with NVR dependencies
                self.dependency.complete_name == other.complete_name
                and self._is_direct_reference(self.dependency)
                != self._is_direct_reference(other)
            )
        )
```

Reply:
Ack. 

NVR = Name, Version, Release. Old habbits. :)